File: Medium/python/2487. Remove Nodes From Linked List.py
"""
You are given the head of a linked list.

Remove every node which has a node with a greater value anywhere to the right side of it.

Return the head of the modified linked list.
"""

from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNodes(self, head: Optional[ListNode]) -> Optional[ListNode]:
        stack = []
        current = head

        while current:
            if stack and stack[-1].val < current.val:
                while stack and stack[-1].val < current.val:
                    stack.pop()
                if stack:
                    stack[-1].next = current

            stack.append(current)
            current = current.next

        return stack[0]

File: Medium/Javascript/2487. Remove Nodes From Linked List.js
/**You are given the head of a linked list.

Remove every node which has a node with a greater value anywhere to the right side of it.

Return the head of the modified linked list. */

class ListNode {
  constructor(val, next) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var removeNodes = function (head) {
  let stack = [];
  let curr = head;

  while (curr) {
    if (stack.length > 0 && stack[stack.length - 1].val < curr.val) {
      while (stack.length > 0 && stack[stack.length - 1].val < curr.val) {
        stack.pop();
      }

      if (stack.length > 0) {
        stack[stack.length - 1].next = curr;
      }
    }

    stack.push(curr);
    curr = curr.next;
  }

  return stack[0];
};

let head = new ListNode(5, new ListNode(2, new ListNode(13, new ListNode(3, new ListNode(8)))))

console.log(removeNodes(head))